app/kudos/utils.py

```diff
@@ -0,0 +1,600 @@
+        kudos = self.getKudosById(kudos_id, to_dict=True)
+        kudos['owner_address'] = self._contract.functions.ownerOf(kudos_id).call()
+        if Token.objects.filter(pk=kudos_id).exists():
+            kudos_token = Token.objects.get(pk=kudos_id)
+            kudos['txid'] = kudos_token.txid
+            updated_kudos_token = Token(pk=kudos_id, **kudos)
+            updated_kudos_token.save()
+        else:
+            kudos_token = Token(pk=kudos_id, **kudos)
+            kudos_token.save()
+        logger.info(f'Synced id #{kudos_token.id}, "{kudos_token.name}" kudos to the database.')
+
+    # @retry
+    def sync_db(self, kudos_id, txid):
+        """Sync up the Kudos contract on the blockchain with the database.
+
+        Args:
+            kudos_id (int): Kudos Id
+            txid (str): The transaction hash.
+
+        Returns:
+            TYPE: Description
+        """
+
+        # Handle the dummy Kudos
+        if kudos_id == 0:
+            return False
+
+        # Grab the Kudos from the blockchain, augment with owner_address
+        kudos = self.getKudosById(kudos_id, to_dict=True)
+        kudos['owner_address'] = self._contract.functions.ownerOf(kudos_id).call()
+
+        # # Update an existing Kudos in the database
+        # if Token.objects.filter(pk=kudos_id).exists():
+        #     if txid:
+        #         kudos['txid'] = txid
+        #     kudos_token = Token(pk=kudos_id, **kudos)
+        #     kudos_token.save(update_fields=list(kudos.keys()))
+        # # Add a new Kudos to the database.  Require txid so we can link to kudos_transfer table.
+        # else:
+        #     if not txid:
+        #         raise ValueError('Must provide a txid when syncing a new Kudos.')
+        kudos['txid'] = txid
+        kudos_token = Token(pk=kudos_id, **kudos)
+        kudos_token.save()
+        # Find the object which matches the kudos that was just cloned
+        try:
+            kudos_transfer = KudosTransfer.objects.get(receive_txid=txid)
+        except KudosTransfer.DoesNotExist:
+            # Only warn for a Kudos that is cloned/transfered, not a Gen0 Kudos.
+            if kudos_token.num_clones_allowed == 0:
+                logger.warning(f'No KudosTransfer object found for Kudos ID {kudos_id}')
+                # raise KudosTransferNotFound(kudos_id, 'No KudosTransfer object found')
+                # raise
+        else:
+            # Store the foreign key reference if the kudos_transfer object exists
+            kudos_transfer.kudos_token = kudos_token
+            kudos_transfer.save()
+
+        logger.info(f'Synced id #{kudos_token.id}, "{kudos_token.name}" kudos to the database.')
+
+    def _get_contract_address(self):
+        """Get the Kudos contract address, depending on the network.
+
+        Returns:
+            str: Kudos contract address.
+        """
+        if self.network == 'mainnet':
+            return to_checksum_address('0x56c72cda0b04fc39a25d0b6a64fa258fad46d664')
```

Comment:
Should we define these contract addresses somewhere else?

Reply:
Yes I think we should.  For Bounties we are also defining it in the code.  Would probably be better suited to `settings.py`.  What do you think?

Reply:
Would be nice to move this out of `utils.py` as well.

https://github.com/jasonrhaas/web/blob/kudos-v1/app/dashboard/utils.py#L221-L226

Reply:
Did this for kudos.  Didn't touch Bounties as its not part of this PR.